cumulusci/tasks/apex/batch.py

```diff
@@ -32,83 +34,116 @@ def _run_task(self):
 
         self.logger.info("Job is complete.")
 
-        vals = {
-            key: value
-            for key, value in self.batch.items()
-            if key
-            in {
-                "Id",
-                "Status",
-                "ExtendedStatus",
-                "NumberOfErrors",
-                "JobItemsProcessed",
-                "TotalJobItems",
-            }
-        }
-        if not self.success:
+        summary = self.summarize_batches(self.batches)
+        failed_batches = self.failed_batches(self.batches)
+
+        if failed_batches:
             self.logger.info("There have been some batch failures.")
             self.logger.info("Error values:")
-            self.logger.info(repr(vals))
-            raise SalesforceException("There were import errors: %s" % repr(vals))
-        elif not self.done_for_sure:
+            self.logger.info(repr(failed_batches))
+            raise SalesforceException(
+                f"There were batch errors: {repr(failed_batches)}"
+            )
+        elif not summary["CountsAddUp"]:
             self.logger.info("The final record counts do not add up.")
             self.logger.info("This is probably related to W-1132237")
-            self.logger.info(repr(vals))
+            self.logger.info(repr(summary))
 
         self.logger.info(
-            "%s took %d seconds to process %d batches.",
-            self.batch["ApexClass"]["Name"],
-            self.delta,
-            self.batch["TotalJobItems"],
+            f"{self.options['class_name']} took {summary['ElapsedTime']} seconds to process {summary['TotalJobItems']} batches."
         )
 
-        return self.success
+    def failed_batches(self, batches: Sequence[dict]):
+        failed_batches = []
+        for batch in batches:
+            if batch["NumberOfErrors"]:
+                failed_batches.append(
+                    {
+                        key: value
+                        for key, value in batch.items()
+                        if key
+                        in {
+                            "Id",
+                            "Status",
+                            "ExtendedStatus",
+                            "NumberOfErrors",
+                            "JobItemsProcessed",
+                            "TotalJobItems",
+                        }
+                    }
+                )
+        return failed_batches
 
     def _poll_action(self):
         # get batch status
-        query_results = self.tooling.query(self._batch_query)
 
-        self.batch = query_results["records"][0]
+        if not self.original_created_date:
+            query_results = self.tooling.query(self._batch_query(date_limit=None))
+            self.original_created_date = parse_api_datetime(
+                query_results["records"][0]["CreatedDate"]
+            )
```

Comment:
We should handle the case where no job was found and show something more helpful than an IndexError.

Reply:
e09332e6d0185b0c2ab675c37c5ea0b232103204